src/Abp.Zero.EntityFramework/EntityHistory/EntityHistoryHelper.cs

```diff
@@ -487,54 +478,100 @@ private void UpdateChangeSet(DbContext context, EntityChangeSet changeSet)
-
-                            continue;
+                            // Update foreign key
+                            propertyChange.NewValue = newValue;
                         }
+                    }
+                }
+
+                /* Update tracked properties */
+                var auditedPropertyChanges = entityChange.PropertyChanges
+                                             .Where(pc => !foreignKeyPropertyNames.Contains(pc.PropertyName))
+                                             .Distinct()
+                                             .ToList();
 
-                        if (propertyChange.OriginalValue == propertyChange.NewValue)
+                foreach (var propertyChange in auditedPropertyChanges)
+                {
+                    var memberEntry = entityEntry.Member(propertyChange.PropertyName);
+                    if (!(memberEntry is DbPropertyEntry))
+                    {
+                        // skipping other types of properties
+                        // - Reference navigation properties (DbReferenceEntry)
+                        // - Collection navigation properties (DbCollectionEntry)
+                        continue;
+                    }
+
+                    var propertyEntry = entityEntry.Property(propertyChange.PropertyName);
+                    var propertyInfo = entityEntry.GetPropertyInfo(propertyChange.PropertyName);
+
+                    // TODO: fix new value comparison before truncation
+                    var newValue = propertyEntry.GetNewValue()?.ToJsonString().TruncateWithPostfix(EntityPropertyChange.MaxValueLength);
+
+                    if (propertyChange.OriginalValue == propertyChange.NewValue)
+                    {
+                        if (propertyChange.NewValue == newValue)
                         {
-                            var newValue = propertyEntry.GetNewValue().ToJsonString();
-                            if (newValue == propertyChange.NewValue)
-                            {
-                                // No change
-                                entityChange.PropertyChanges.Remove(propertyChange);
-                            }
-                            else
-                            {
-                                // Update foreign key
-                                propertyChange.NewValue = newValue.TruncateWithPostfix(EntityPropertyChange.MaxValueLength);
-                            }
+                            // No change
+                            entityChange.PropertyChanges.Remove(propertyChange);
+                        }
+                        else
+                        {
+                            // Update audited property
+                            propertyChange.NewValue = newValue;
                         }
                     }
                 }
             }
         }
+
+        private EntityPropertyChange CreateEntityPropertyChange(object oldValue, object newValue, PropertyInfo propertyInfo)
```

Comment:
Can accept `propertyEntry` since `propertyInfo` is only used within this private function?

Reply:
Nope. `GetPropetryChanges` uses `PropertyEntry` to get new/old values but `GetRelationshipChanges` uses `Tuple<EntityState.Name, ObjectStateEntry.State, EntityKey>` to get new/old values.